internal/dynamicwatcher/dynamic_watcher.go

```diff
@@ -0,0 +1,305 @@
+package dynamicwatcher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	ctrlcache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// OwnerReference points to a single owner of an watch operation.
+type OwnerReference struct {
+	schema.GroupKind
+	UID       types.UID
+	Name      string
+	Namespace string
+}
+
+// Namespaced GroupVersionKind
+type NamespacedGKV struct {
+	schema.GroupVersionKind
+	Namespace string
+}
+
+// usually fulfilled by meta.RESTMapper
+type restMapper interface {
+	RESTMapping(gk schema.GroupKind, versions ...string) (
+		*meta.RESTMapping, error)
+}
+
+// DynamicWatcher is able to dynamically allocate new watches for arbitrary objects.
+// Multiple watches to the same resource, will be de-duplicated.
+type DynamicWatcher struct {
+	log        logr.Logger
+	scheme     *runtime.Scheme
+	restMapper restMapper
+	client     dynamic.Interface
+
+	opts DynamicWatcherOptions
+
+	sinksLock sync.RWMutex
+	sinks     []watchSink
+
+	informersLock      sync.Mutex
+	informersStopCh    map[NamespacedGKV]chan<- struct{}
+	informerReferences map[NamespacedGKV]map[OwnerReference]struct{}
+}
+
+const defaultEventHandlerResyncPeriod = 10 * time.Hour
+
+type informer interface {
```

Comment:
Just a question: Why not use one of the informers from the `client-go` package? https://pkg.go.dev/k8s.io/client-go/informers